File: design_patterns/join/join_java/src/main/java/app/swarch/designpatterns/join/JoinDemo.java
package app.swarch.designpatterns.join;


import java.util.*;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.function.Function;
import java.util.stream.Stream;

/**
 * Reply received from a Channel, after submitting the message.
 * For Asynchronous Channels it is Reply<Void>.
 *
 * @param <T> - value sent within the reply.
 */
interface Reply<T> {

    T get();
}

/**
 * Message sent to a Channel.
 *
 * @param <T> - value sent within the message.
 */
interface Message<T> {

    T get();
}

/**
 * Channel, which accepts the Messages and responds with Replies.
 *
 * @param <M> - Message value type
 * @param <R> - Reply value type
 */
abstract class Channel<M, R> {

    // Monitor object for the queue to implement kind of a blocking-empty check.
    private final Object queueMonitor = new Object();
    // Blocking queue for incoming messages, which are popped when particular JoinPattern matches,
    // which means all other channels joined in that pattern also have incoming messages
    protected BlockingQueue<Message<M>> messageQueue = new LinkedBlockingQueue<>();

    /**
     * Method of sending messages to the Channel.
     *
     * @param message - submitted by a client and can carry a specific value or Void.
     * @return reply - Asynchronous Channels return immediately Reply<Void>. Synchronous Channels block until
     *      reply is provided through {@link #provideReply(Reply)}.
     * @throws InterruptedException - can be thrown by the Synchronous Channels, since they are blocking the thread
     *      until reply is provided.
     */
    public abstract Reply<R> send(Message<M> message) throws InterruptedException;

    /**
     * When Join Pattern fully matched and got executed by Scheduler, it provides the execution result as a reply
     * to the waiting channel, so that channel can deliver the reply to the calling client.
     *
     * @param reply - execution result from Join Pattern
     * @throws InterruptedException - can be thrown by Synchronous Channels, since they are putting Reply to the queue,
     *    so that it can be provided to the client in the waiting thread.
     */
    public void provideReply(Reply<R> reply) throws InterruptedException {
    }

    /**
     * Blocking method to wait until there are changes in the queue. Join Pattern cannot poll the queues of its
     * Channels, because even if all the Channels do have messages, the pattern may not be necessarily executed.
     * It depends on the Scheduler's Pattern Selection Policy - there can be several matching Patterns with the same
     * Channels.
     *
     * @return true, if Channel's queue has messages.
     * @throws InterruptedException - the method uses waiting for monitor object, which can produce the exception
     */
    public boolean waitUntilHasMessages() throws InterruptedException {
        if (!messageQueue.isEmpty()) {
            return true;
        }
        synchronized (queueMonitor) {
            if (!messageQueue.isEmpty()) {
                return true;
            }
            queueMonitor.wait();
        }
        return !messageQueue.isEmpty();
    }

    /**
     * Non-blocking method to check if Channel's queue has messages. Returns result immediately. It is used by the
     * Scheduler to select all matching Join Patterns.
     *
     * @return true, if Channel's queue has messages.
     */
    public boolean hasMessages() {
        return !messageQueue.isEmpty();
    }

    /**
     * Blocking operation to retrieve and remove the head of the Channel's Message queue.
     * @return the head Message from the queue.
     * @throws InterruptedException - blocking queue may produce the exception
     */
    public Message<M> getMessage() throws InterruptedException {
        return messageQueue.take();
    }

    /**
     * Puts the Message into the queue and notifies Monitor.
     * @param message - new sent Message
     * @throws InterruptedException - putting into the blocking queue may produce the exception
     */
    protected void processMessage(Message<M> message) throws InterruptedException {
        messageQueue.put(message);
        synchronized (queueMonitor) {
            queueMonitor.notifyAll();
        }
    }
}

/**
 * Synchronous Channel blocks when sending the Message until Reply is available.
 *
 * @param <M> - Message value type
 * @param <R> - Reply value type
 */
class SynchronousChannel<M, R> extends Channel<M, R>{

    // Queue for the provided replies
    protected BlockingQueue<Reply<R>> replyQueue = new LinkedBlockingQueue<>();

    @Override
    public void provideReply(Reply<R> reply) throws InterruptedException {
        replyQueue.put(reply);
    }

    @Override
    public Reply<R> send(Message<M> message) throws InterruptedException {
        processMessage(message);
        return replyQueue.take();
    }
}

/**
 * Asynchronous Channel doesn't block when sending the Message. It responds with Reply<Void> immediately.
 *
 * @param <M> - Message value type
 */
class AsynchronousChannel<M> extends Channel<M, Void> {

    @Override
    public Reply<Void> send(Message<M> message) {
        new Thread(() -> {
            try {
                processMessage(message);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }).start();
        return () -> null;
    }
}

/**
 * Join definition of several channels. It is expected to have at leas 1 Output Channel.
 * If Output Channel is Synchronous, then Client expects to receive a value after sending a Message to it (a Reply).
 * Output Channel can also be Asynchronous, then Client just expects to execute some asynchronous operation,
 * when Pattern is matched.
 *
 * Other so-called Input Channels are all Asynchronous Channels, which can accept Messages with actual values or just
 * empty Messages (with Void values). In case of empty Messages, such Channels are used as conditions for Pattern
 * matching.
 *
 * @param <T> - return type of the function that is executed when Pattern is matched.
 */
class JoinPattern<T> {

    // Output Channel is referenced separately from other (Input) Channel,
    // so that Pattern can provide the function execution result to it,
    // and consequently to the waiting Client, if Output Channel is Synchronous
    private Channel<?, T> outputChannel;
    // Input Channels are always Asynchronous
    private final List<AsynchronousChannel<?>> inputChannels = new ArrayList<>();
    // Function executed, when Pattern is matched. Result of the execution is provided as a Reply
    // to the Output Channel
    private Function<List<Message<?>>, T> function;

    /**
     * Specify the initial Output Channel.
     *
     * @param outputChannel - initial Channel, which is used as Output of this Pattern execution
     * @return this Join Pattern
     */
    public JoinPattern<T> with(Channel<?, T> outputChannel) {
        Objects.requireNonNull(outputChannel);
        this.outputChannel = outputChannel;
        return this;
    }

    /**
     * Join next Asynchronous Input Channel.
     *
     * @param inputChannel - one of the joined Asynchronous Channels
     * @return this Join Pattern
     */
    public JoinPattern<T> and(AsynchronousChannel<?> inputChannel) {
        Objects.requireNonNull(inputChannel);
        this.inputChannels.add(inputChannel);
        return this;
    }

    /**
     * Specify function, which needs to be executed upon Pattern matching.
     *
     * @param function to execute
     * @return this Join Pattern
     */
    public JoinPattern<T> onMatch(Function<List<Message<?>>, T> function) {
        this.function = function;
        return this;
    }

    /**
     * Blocks until Channels wake up to respond whether they have messages in their queues or not.
     *
     * @throws InterruptedException - can be produced when waiting for Channels
     */
    public void waitUntilMatches() throws InterruptedException {
        if (outputChannel.waitUntilHasMessages()) {
            inputChannels.forEach(channel -> {
                try {
                    channel.waitUntilHasMessages();
                } catch (InterruptedException e) {
                    throw new RuntimeException(e);
                }
            });
        }
    }

    /**
     * Non-blocking method to check whether all Pattern's Channels received Messages.
     * @return true if all Channels have at least one Message in their queues.
     */
    public boolean matches() {
        return outputChannel.hasMessages() && inputChannels.stream().allMatch(Channel::hasMessages);
    }

    /**
     * Executed by Scheduler, if Pattern matches and Scheduler chooses it according to its Selection Policy
     *
     * @throws InterruptedException - providing reply to the Output Channel can cause the exception,
     *      since Synchronous Channels are passing Replies through the blocking queue.
     */
    public void execute() throws InterruptedException {
        List<Message<?>> messages = Stream
            .concat(
                Stream.of(outputChannel.getMessage()),
                Stream.of(
                    inputChannels
                        .stream()
                        .map(channel -> {
                            try {
                                return channel.getMessage();
                            } catch (InterruptedException e) {
                                throw new RuntimeException(e);
                            }
                        })
                        .toArray(Message<?>[]::new)
                )
            )
            .toList();
        // Function execution result is provided to Output Channel as a Reply,
        // so that the Channel can return it to the waiting Client
        outputChannel.provideReply(() -> function.apply(messages));
    }
}

/**
 * Selection Policy helps Scheduler to decide how to choose between multiple matching Patterns.
 */
interface PatternSelectionPolicy {

    /**
     * Select a single Join Pattern from the list of multiple matched Patterns.
     *
     * @param joinPatterns - list of matched Join Patterns
     * @return selected Join Pattern
     */
    JoinPattern<?> select(List<JoinPattern<?>> joinPatterns);
}

/**
 * Selects first occurrence from the ordered list of Patterns.
 */
class FirstOrderedSelectionPolicy implements PatternSelectionPolicy {

    @Override
    public JoinPattern<?> select(List<JoinPattern<?>> joinPatterns) {
        for (JoinPattern<?> joinPattern : joinPatterns) {
            if (joinPattern.matches()) {
                return joinPattern;
            }
        }
        return null;
    }
}

/**
 * Selects first occurrence from the list of Patterns after randomly reshuffling it.
 */
class FirstRandomSelectionPolicy implements PatternSelectionPolicy {

    @Override
    public JoinPattern<?> select(List<JoinPattern<?>> joinPatterns) {
        List<JoinPattern<?>> reshuffledPatterns = new ArrayList<>(joinPatterns);
        Collections.shuffle(reshuffledPatterns);
        for (JoinPattern<?> joinPattern : reshuffledPatterns) {
            if (joinPattern.matches()) {
                return joinPattern;
            }
        }
        return null;
    }
}

/**
 * Runs a loop to check for Patterns that matched (all Patterns' Channels received Messages),
 * and if there are several of them, selects one for execution according to its Selection Policy.
 */
class Scheduler {

    // Selection Policy to decide how to choose from multiple matched Patterns.
    private final PatternSelectionPolicy selectionPolicy;
    // List if provided Join Patterns
    private final List<JoinPattern<?>> joinPatterns;
    // Flag to terminate the loop, when stopped
    private boolean stopped = false;
    // Monitor object for communicating the signals between Join Pattern specific threads
    // and Scheduler's main loop.
    private final Object monitor = new Object();

    public Scheduler(PatternSelectionPolicy selectionPolicy, List<JoinPattern<?>> joinPatterns) {
        this.selectionPolicy = selectionPolicy;
        this.joinPatterns = joinPatterns;
    }

    /**
     * Runs a loop to wait for Patterns that matched, thus all Channels received the messages.
     * Selects one Patterns for execution based on Selection Policy.
     */
    public void run() {
        // Scheduler starts a separate thread for each Pattern, so that
        // main loop can be placed into wait-state. Main loop thread is waken up
        // if at least one Patter-specific thread notifies about matching.
        for (JoinPattern<?> joinPattern : joinPatterns) {
            Thread t = new Thread(() -> {
                while (!stopped) {
                    try {
                        joinPattern.waitUntilMatches();
                        synchronized (monitor) {
                            monitor.notify();
                        }
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            });
            t.setDaemon(true);
            t.start();
        }
        while (!stopped) {
            try {
                synchronized (monitor) {
                    monitor.wait();
                    // Here Scheduler once again retrieves all matched Pattern and then selects one for execution
                    List<JoinPattern<?>> matchingPatterns = joinPatterns.stream().filter(JoinPattern::matches).toList();
                    JoinPattern<?> pattern = selectionPolicy.select(matchingPatterns);
                    if (pattern != null) {
                        pattern.execute();
                    }
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    /**
     * Stops the Scheduler's loop
     */
    public void stop() {
        this.stopped = true;
    }
}



public class JoinDemo {

    public static void main(String[] args) throws Exception {
        SynchronousChannel<String, String> channelA = new SynchronousChannel<>();
        AsynchronousChannel<String> channelB = new AsynchronousChannel<>();
        AsynchronousChannel<String> channelC = new AsynchronousChannel<>();
        AsynchronousChannel<String> channelD = new AsynchronousChannel<>();
        AsynchronousChannel<String> channelE = new AsynchronousChannel<>();
        JoinPattern<String> joinPattern1 = new JoinPattern<String>()
            .with(channelA)
            .and(channelB)
            .and(channelC)
            .onMatch(messages -> "Pattern 1 received messages: " + String.join(",", messages.stream().map(m -> (String) m.get()).toList()));
        JoinPattern<String> joinPattern2 = new JoinPattern<String>()
            .with(channelA)
            .and(channelB)
            .and(channelD)
            .onMatch(messages -> "Pattern 2 received messages: " + String.join(",", messages.stream().map(m -> (String) m.get()).toList()));
        JoinPattern<String> joinPattern3 = new JoinPattern<String>()
            .with(channelA)
            .and(channelB)
            .and(channelE)
            .onMatch(messages -> "Pattern 3 received messages: " + String.join(",", messages.stream().map(m -> (String) m.get()).toList()));
        Scheduler scheduler = new Scheduler(new FirstRandomSelectionPolicy(), List.of(
            joinPattern1,
            joinPattern2,
            joinPattern3
        ));
        Thread t = new Thread(scheduler::run);
        t.setDaemon(true);
        t.start();

        channelB.send(() -> "B");
        channelC.send(() -> "C");
        channelD.send(() -> "D");
        channelE.send(() -> "E");

        System.out.println(channelA.send(() -> "A").get());

        channelB.send(() -> "B");

        System.out.println(channelA.send(() -> "A").get());

        channelB.send(() -> "B");

        System.out.println(channelA.send(() -> "A").get());

        scheduler.stop();
    }
}

File: design_patterns/join/join_pseudo/src/join_demo.py
# A message sent to a channel
class Message:
    value: object

    # constructor accepts message value
    def __init__(self, value):
        self.value = value

    # retrieves the message value
    def get(self) -> object:
        return self.value


# A reply received from the channel
# after message sending
class Reply:
    value: object

    # constructor accepts reply value
    def __init__(self, value):
        self.value = value

    # retrieves the reply value
    def get(self) -> object:
        return self.value


# Queue for messages submitted to a channel
class MessageQueue:
    messages = []

    # get the head of the message queue
    def get(self) -> Message:
        return self.messages.pop()

    # add new message to the queue
    def put(self, message: Message):
        self.messages.append(message)

    # check if queue is empty
    def isEmpty(self) -> bool:
        return len(self.messages) > 0


# Queue for replies provided to a channel
# by pattern, when executing its function
class ReplyQueue:
    replies = []

    # get the head of the reply queue
    def get(self) -> Reply:
        return self.replies.pop()

    # add new reply to the queue
    def put(self, reply: Reply):
        self.replies.append(reply)

    # check if queue is empty
    def isEmpty(self) -> bool:
        return len(self.replies) > 0


# Accepts messages and returns replies
class Channel:
    messageQueue: MessageQueue = MessageQueue()

    # Send a message to the channel
    # By default (as in case with async channels)
    # replies immediate with empty value
    def send(self, message: Message) -> Reply:
        self.messageQueue.put(message)
        return Reply(None)

    # Reply provided by a pattern after it executed
    # its function
    def provideReply(self, reply: Reply):
        # by default does nothing,
        # replies are relevant for sync channels
        pass

    # Checks if the channel has messages to process
    def hasMessage(self) -> bool:
        return self.messageQueue.isEmpty() is False


# Sync channel blocks the caller until reply is available
class SynchronousChannel(Channel):
    replyQueue: ReplyQueue = ReplyQueue()

    # Send a message to the channel
    # Blocks until reply is available
    def send(self, message: Message) -> Reply:
        self.messageQueue.put(message)
        # in actual implementation getting reply
        # from the queue would be a blocking operation
        return Reply(self.replyQueue.get())

    # Reply provided by a pattern after it executed
    # its function
    def provideReply(self, reply: Reply):
        self.replyQueue.put(reply)


# Async channel does not block when sending a message
# It replies immediate with empty value
class AsynchronousChannel(Channel):
    pass


# Join definition combining several channels.
# Join pattern matches, when all its combined channels
# have at least 1 message to process
class JoinPattern:
    outputChannel: Channel
    inputChannels = []
    function: any

    # Join initial channel, which will be provided
    # with the result of pattern's function execution
    def when(self, channel: Channel):
        self.outputChannel = channel
        return self

    # Join async input channels
    def andAlso(self, channel: AsynchronousChannel):
        self.inputChannels.append(channel)
        return self

    # Provide a function to execute, when pattern matches
    def onMatch(self, function):
        self.function = function
        return self

    # Check if pattern matches
    def matches(self) -> bool:
        return self.outputChannel.hasMessage() \
               and all([inputChannel.hasMessage()
                        for inputChannel in self.inputChannels])

    # Execute pattern's function
    def execute(self):
        result = self.function()
        self.outputChannel.provideReply(Reply(result))


# Selection policy helps Scheduler to decide
# how to choose between multiple matching patterns.
class PatternSelectionPolicy:

    def select(self, patterns: []) -> JoinPattern:
        pass


# Select the first element in the provided list of patterns
class ConcreteSelectionPolicy(PatternSelectionPolicy):

    def select(self, patterns: []) -> JoinPattern:
        # select first element
        return patterns[0]


# Executes a 'runnable' function in a separate thread
class Thread:

    # Start the thread
    def start(self, runnable):
        # runs in a separate thread
        runnable()


# Runs a loop to check for patterns that matched
# (all pattern's channels received messages),
# and if there are several of them, selects one
# for execution according to its selection policy.
class Scheduler:
    joinPatterns = []
    selectionPolicy: PatternSelectionPolicy

    def __init__(self,
                 joinPatterns: [],
                 selectionPolicy: PatternSelectionPolicy):
        self.joinPatterns = joinPatterns
        self.selectionPolicy = selectionPolicy

    # Runs a loop to wait for Patterns that matched
    def run(self):
        def runnable():
            while True:
                matchingPatterns = [pattern
                                    for pattern
                                    in self.joinPatterns
                                    if pattern.matches()]
                selectedPattern = self.selectionPolicy\
                    .select(matchingPatterns)
                selectedPattern.execute()

        Thread().start(runnable)

# This is a Python-based pseudocode written with the intent
# to provide intuition about the join-calculus pattern.
# It is not a real runnable code
if __name__ == '__main__':
    # create 4 channels
    channel_a = SynchronousChannel()
    channel_b = AsynchronousChannel()
    channel_c = AsynchronousChannel()
    channel_d = AsynchronousChannel()
    # create 2 patterns with same output channel
    # and only 1 difference in input channels
    def function_1():
        return "1"
    pattern_1 = JoinPattern() \
        .when(channel_a) \
        .andAlso(channel_b) \
        .andAlso(channel_c) \
        .onMatch(function_1)
    def function_2():
        return "2"
    pattern_2 = JoinPattern() \
        .when(channel_a) \
        .andAlso(channel_b) \
        .andAlso(channel_d) \
        .onMatch(function_2)
    scheduler = Scheduler([pattern_1, pattern_2], ConcreteSelectionPolicy())
    scheduler.run()

    # here we first send messages to async input channels
    # so that when we send later message to sync channel A
    # scheduler is forced to choose between 2 patterns
    channel_b.send(Message("B"))
    channel_c.send(Message("C"))
    channel_d.send(Message("D"))

    # message to channel A causes both patterns (1 and 2)
    # to match, because all the other of their channels
    # also received messages. It is up to scheduler and
    # its selection policy to decide, which pattern to select
    # for execution
    reply = channel_a.send(Message("A"))

    # here we expect result to be "1"
    # because when choosing between patterns 1 and 2,
    # scheduler's policy selects the first match
    print(reply.get())
